drivers: flash: stm32 qspi: Make delayed data sampling configurable

adazot opened the pull request:
PR #92742 enabled STM32 QSPI delayed data sampling (SSHIFT) for single flash mode. This caused communication issues with our flash chip starting with SFDP reading. From my point of view this is regression introduced in v4.3.

The only reliable solution found was to disable SSHIFT. SSHIFT is not configurable in the current codebase.

This PR implements configurable SSHIFT (ssht-enable), similarly to what XSPI and OSPI drivers implement.

Fixes #99099

adazot committed:
drivers: flash: stm32 qspi: Make delayed data sampling configurable

The QSPI delayed data sampling (SSHIFT) is enabled by default. This
feature is configurable in both XSPI and OSPI drivers. Align with
these drivers and make the feature configurable for QSPI too.

Signed-off-by: Adam Zotow <azo@trackunit.com>

adazot committed:
dts: bindings: stm32 qspi: add ssht-enable property

This property allows to enable QSPI delayed data sampling (SSHIFT).

Signed-off-by: Adam Zotow <azo@trackunit.com>

JarmouniA reviewed (dismissed):
The original behavior was that Sample Shifting is enabled on all platforms using dual-flash. Then it got enabled for all platforms https://github.com/zephyrproject-rtos/zephyr/commit/b147d78c2f183c6ded5342e77fe8af5d97433779. Now you are disabling it on all platforms, which is not right.
You need to add the new property to all in-tree platforms using dual-flash to restore the original setting.

JarmouniA commented:
Cc @taltenbach

adazot committed:
boards: stm32: set ssht-enable property on all qspi dual-flash boards

With ssht-enable property introduced in STM32 QSPI driver the delayed
data sampling (SSHIFT) got disabled for both dual and non-dual flash
configurations. Set ssht-enable property for dual flash boards
to reenable delayed data sampling.

Signed-off-by: Adam Zotow <azo@trackunit.com>

etienne-lms reviewed (commented):
I think it would be more consistent to place commit *"dts: bindings: stm32 qspi: add ssht-enable property"* before commit *"drivers: flash: stm32 qspi: Make delayed data sampling configurable"* in your series.
That said, LGTM.

JarmouniA commented:
> You need to add the new property to all in-tree platforms using dual-flash to restore the original setting.

This is also not sufficient. Downstream users, with dual-flash setups, who relied on the feature being enabled by default, will have to add the new property in DT. So an entry in the migration guide is needed.

adazot committed:
doc: migration-guide-4.4: Add st,stm32-qspi ssht-enable note

Add migration note for dual-flash configurations requiring ssht-enable
property to reenable sample shifting.

Signed-off-by: Adam Zotow <azo@trackunit.com>

etienne-lms commented on doc/releases/migration-guide-4.3.rst:
```diff
@@ -167,6 +167,14 @@ Phy
+QSPI
+===
+
+* :dtcompatible:`st,stm32-qspi` compatible nodes configured with ``dual-flash`` property
```
Is it really only the dual-flash devices that need to set `ssht-enable`?

etienne-lms reviewed (commented):
It seems there are other boards that need to be updated with

azo-trackunit commented on dts/bindings/qspi/st,stm32-qspi.yaml:
```diff
@@ -59,6 +59,12 @@ properties:
 
       For example
          dma-names = "tx_rx";
+  ssht-enable:
```
missed that, thanks
